src/Cache.ts

```diff
@@ -0,0 +1,68 @@
+/**
+ * Standard cache interface that allows for the storage (store) and retrieval (lookup) of
+ * key-value pairs with false negatives: The lookup of a key might return undefined even
+ * if the key was previously stored, but if the lookup does return a defined value, the value is
+ * guaranteed to be that of the most recent store for the key.
+ */
+export interface Cache<KeyType, ValueType> {
+
+  /**
+   * Store a key-value pair in the cache.
+   */
+  store (key: KeyType, value: ValueType): void;
+
+  /**
+   * Lookup the value for a specified key with possible false negatives. See comment
+   * above the interface definition.
+   */
+  lookup (key: KeyType): ValueType | undefined;
+}
+
+/**
+ * A simple associative cache that stores the cached objects (key-value pairs)
+ * in a fixed size array along with a hashmap indexing the position of each
+ * stored key. The current implementation uses a simple random replacement
```

Comment:
Consider a equally simple (if not simpler) but more effective (in the initial phase where all slots will guaranteed to be used.) scheme/implementation: first in first out/circular cache.

Reply:
Interestingly, the random strategy has provably good performance. That said, I think a good strategy would be to use something like a clock algorithm used in OS caches. I will transition it to that strategy sometime when you are away. For now, I will keep the random strategy - the random strategy should not be much worse than clock and we should get decent enough performance with large cache sizes (as you correctly point out).

Reply:
Sounds good.